### ISSUE TITLE
Update phpdocumentor/type-resolver version in PHPUnit phar

### ISSUE THREAD

**Issue by nikic** [sebastianbergmann/phpunit#5712]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 9.6.16
| PHP version         | 7.4.21
| Installation Method | PHAR

#### Summary

Based on the composer.lock file in the phar, it currently contains `phpdocumentor/type-resolver` version 1.6.1, which is quite old (predating the release date of PHPUnit 9.6.16 significantly). In particular, it does not contain the fix for PHP-Parser 5 compatibility in https://github.com/phpDocumentor/TypeResolver/pull/197, which is part of 1.7.4. This caused a good bit of confusion over in https://github.com/nikic/PHP-Parser/issues/981, because an error was only reproducible with the phpunit phar release, but not the composer dev dependency.

Is it possible to update this dependency? It's not really obvious to me how this version is determined, because the composer.lock in the phar doesn't seem to match https://github.com/sebastianbergmann/phpunit/blob/9.6/composer.lock at all.

**Comment by llaville** [comment]
Hello @sebastianbergmann 

As author of https://github.com/nikic/PHP-Parser/issues/981, I approve this request, especially because it has also an impact on GitHub Action `shivammathur/setup-php@v2`.

This action uses the PHAR version of PHPUnit (see https://github.com/shivammathur/setup-php/blob/v2/dist/index.js#L969)

So, when we run a workflow with such declaration : 

```yaml
      uses: shivammathur/setup-php@v2
        with:
          php-version: 7.4
          tools: phpunit:9.6
```
We've got trouble ! For example : https://github.com/composer-unused/symbol-parser/actions/runs/7990344120/job/21818955371#step:5:15

Workaround, as [suggested Nikita](https://github.com/nikic/PHP-Parser/issues/981#issuecomment-1958016088), is to use the `vendor/bin/phpunit` (installed with Composer).

And it's works ! See https://github.com/composer-unused/symbol-parser/actions/runs/7996055440

**Comment by sebastianbergmann** [comment]
> It's not really obvious to me how this version is determined, because the composer.lock in the phar doesn't seem to match

With PHPUnit 9.6, the `composer.lock` file that is in the `9.6` branch of this repository is not build `phpunit.phar`. When building the PHAR, additional packages are installed, which results in a modified `composer.lock`. It is this modified `composer.lock` that you see when you introspect the PHAR like so:

```
$ wget https://phar.phpunit.de/phpunit-9.6.phar
$ php phpunit-9.6.phar --composer-lock
```

For [backward compatibility](https://docs.phpunit.de/en/9.6/test-doubles.html#prophecy), the PHAR build process for PHPUnit 9.6 installs `phpspec/prophecy` with version constraint `^1.12.1`. This depends on `phpdocumentor/reflection-docblock` which in turn depends on `phpdocumentor/type-resolver`.

> Is it possible to update this dependency?

I do not know how. Let me explain:

PHPUnit's `composer.json` file on the `9.6` branch uses

```json
"config": {
    "platform": {
        "php": "7.3.0"
```

to make sure that no package requiring more than PHP 7.3.0 can be installed regardless of the actual PHP version that is used to build the PHAR, for instance. This leads to version 1.6.1 of `phpdocumentor/type-resolver` to be installed. When I change `"php": "7.3.0"` to `"php": "8.3.0"` in the above, then the PHAR build process uses version 1.8.1 of `phpdocumentor/type-resolver`.

This is due to the fact that [version 1.6.1 of `phpdocumentor/type-resolver` uses the `"php": "^7.2 || ^8.0"` version constraint](https://github.com/phpDocumentor/TypeResolver/blob/1.6.1/composer.json#L13) where [version 1.8.1 of `phpdocumentor/type-resolver` uses the `"php": "^7.4 || ^8.0"` version constraint](https://github.com/phpDocumentor/TypeResolver/blob/1.8.1/composer.json).

**Comment by jrfnl** [comment]
@jaapio Any chance of TypeResolver supporting PHP 7.3 ?

**Comment by nikic** [comment]
@sebastianbergmann Thank you the detailed explanation!

The easiest solution would probably be to release a 1.6.2 version of phpdocumentor/type-resolver with that patch cherry-picked. But maybe this isn't worth the bother, given that using phpunit as a dev dependency avoids the problem.

**Comment by jaapio** [comment]
I think we can fix that quite easy, will investigate that tomorrow and create a new release 1.6.2

**Comment by jaapio** [comment]
I managed to have a look today, what would be the easiest way for me to validate this issue is solved before I create a new release?

**Comment by sebastianbergmann** [comment]
> what would be the easiest way for me to validate this issue is solved before I create a new release?

I do not how this could be validated before the release. Once that release has been made, you can use `ant phar` on PHPUnit's `9.6` branch. This should then use the new release.

**Comment by jaapio** [comment]
Ok, will give it a try

**Comment by sebastianbergmann** [comment]
> will investigate that tomorrow and create a new release 1.6.2

[1.6.2](https://github.com/phpDocumentor/TypeResolver/releases/tag/1.6.2) was already tagged in October 2022. We are talking about 1.6.3 as the new version here, right?

**Comment by nikic** [comment]
Oh, that's tricky. I didn't realize that the PHP 7.4 requirement was added in 1.6.2, rather than 1.7.0. In that case the version would have to be something like `1.6.1-p1` I think? (Based on looking at https://getcomposer.org/doc/04-schema.md#version, though it's not explicit about how the versions sort there).

**Comment by llaville** [comment]
@nikic If we follow real semantic perharps. But for an older version I think we may accept to downgrade PHP compatibility between 1.6.2 (`php: ^7.4 || ^8.0`)  and 1.6.3 (`php: ^7.3 || ^8.0`)
Look at download stats https://packagist.org/packages/phpdocumentor/type-resolver/stats#1.6.2

**Comment by jaapio** [comment]
PR is ready, please let me know what you think, apart from 7.3 drop there is nothing breaking, yes we had to remove all typed properties. But that's about it

**Comment by llaville** [comment]
@jaapio I 've checked your PR and check your new branch `php73` locally, rebuilding a PHPUnit 9.6 phar locally too !


I've re-tested again the https://github.com/composer-unused/symbol-parser project on PHP 7.4 platform

```
PHP 7.4.33 (cli) (built: Nov 15 2022 06:05:55) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.33, Copyright (c), by Zend Technologies
    with Xdebug v3.1.6, Copyright (c) 2002-2022, by Derick Rethans
```

With `nikic/php-parser v5.0.1  requires php (>=7.4)` installed.

With a rebuild of PHPUnit 9.6 with code of https://github.com/phpDocumentor/TypeResolver/tree/php73 seems works fine !


```
devilbox@php-7.4.33 in /shared/backups/forks/symbol-parser $ /shared/backups/github/phpunit/phpunit.phar --bootstrap phpunit.bootstrap.php --no-configuration -v tests/
PHPUnit 9.6.16 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.33

........................................................          56 / 56 (100%)

Time: 00:00.101, Memory: 20.00 MB

OK (56 tests, 105 assertions)
```

**Comment by jaapio** [comment]
Thanks for confirming, I will merge this and tag a new version

**Comment by llaville** [comment]
> Thanks for confirming, I will merge this and tag a new version

Hope I don't missed anything ;-)

**Comment by llaville** [comment]
@jaapio Is it not supposed to release a 1.6.3 rather than https://github.com/phpDocumentor/TypeResolver/releases/tag/1.8.2 ?

**Comment by jaapio** [comment]
No, because the base version was 1.8.1, the 1.6 series is not compatible with the new php parser, it would have resulted in the same errors.

**Comment by sebastianbergmann** [comment]
I just noticed that with `phpdocumentor/type-resolver` 1.8.2 two new dependencies are pulled in: `doctrine/deprecations` and `phpstan/phpdoc-parser`:
```patch
-phpunit/phpunit: 9.6.16
+phpunit/phpunit: 9.6@b59fa6534869a57ccfe0142d08aef3db59cd2680
+doctrine/deprecations: 1.1.3
 doctrine/instantiator: 1.5.0
 myclabs/deep-copy: 1.11.1
 nikic/php-parser: v4.18.0
@@ -6,8 +7,9 @@
 phar-io/version: 3.2.1
 phpdocumentor/reflection-common: 2.2.0
 phpdocumentor/reflection-docblock: 5.3.0
-phpdocumentor/type-resolver: 1.6.1
+phpdocumentor/type-resolver: 1.8.2
 phpspec/prophecy: v1.18.0
+phpstan/phpdoc-parser: 1.25.0
 phpunit/php-code-coverage: 9.2.30
 phpunit/php-file-iterator: 3.0.6
 phpunit/php-invoker: 3.1.1
```

**Comment by sebastianbergmann** [comment]
I have released [PHPUnit 9.6.17](https://github.com/sebastianbergmann/phpunit/blob/9.6/ChangeLog-9.6.md#9617---2024-02-23). The PHAR for this release has the updated dependencies.

**Comment by llaville** [comment]
I confirmed that all still run fine with official PHAR release 9.6.17 on PHP 7.4 and 8.2 (my quick tests about contextual project https://github.com/composer-unused/symbol-parser/ that used nikic/php-parser 4.18 or 5.0)
